### PR TITLE
Whitelist enhancement

### PR DESCRIPTION
**Pull request by lanzone-myghty**
Added the possibility to use patterns for includes and excludes list.

Motivation:

in bigger projects it may be require to add a whole package to the includes and exludes list. As right now string comparison is used this is not possible.

Modifications:

Enhanced the resolveSimulations method that it is capable of dealing with patterns like somepackage.* which will include then all classes in that package instead of only being able to explicitly define the classes. The old way will still work.

Result:

Developers are now able to define whole packages using a star at the end instead of defining the whole class name.

### CODE REVIEW

**Comment by slandelle** [comment]
Hi @lanzone-myghty 

Thanks for contributing!

Actually, there are some pieces missing. Could you please
* sign our [CLA](https://docs.google.com/a/gatling.io/forms/d/1hbpxVqJ5hIYYJuOjOyOzub9fJFGFGvvfnB5w5LCWdUw)
* add JUnit tests proving that:
  * your new feature indeed works
  * you didn't introduce a regression or a breaking change causing existing setups to break with your changes
* [Document the new feature](https://github.com/gatling/gatling/blob/master/src/sphinx/extensions/maven_plugin.rst) (would be a separate PR as documentation lives in main repository)

Cheers :)

**Comment by lanzone-myghty** [comment]
Hi @slandelle , 

CLA is signed. But JUnit Tests? The Plugin does not even have a Test-Source Folder. I can add JUnit tests, but that will be way more work, as the mojo is not designed for unit testing which means I will have to use either reflection or introduce new public methods or change the method visibility. It will also introduce dependencies for mocking. What do you prefer?

**Comment by slandelle** [comment]
Yes, please add a src/test/java directory and a JUnit dependency with test scope. You can extract your filtering logic to some helper.

**Comment by lanzone-myghty** [comment]
will do that :)

**Comment by slandelle** [comment]
Thanks :)

**Comment by lanzone-myghty** [comment]
Added some unit tests. Not yet ideal, but at the moment my time is very limited :/ sorry for that @slandelle

**Comment by lanzone-myghty** [comment]
Hey @slandelle :) Thanks for the comments, I will go through them as soon as I have time, will be busy for the next two weeks as end of project is approaching but afterwards I will take care of them!

File: gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java [hunk 1]
```diff
@@ -404,6 +400,57 @@ public boolean accept(File pathname) {
...
+      String regexString = patternString.replace(".", "\\.");
+      regexString = regexString.replace("*", "([a-zA-Z\\.]*)");
+      includes.add(Pattern.compile(regexString));
+    });
+
+    MojoUtils.arrayAsListEmptyIfNull(this.excludes).forEach(patternString -> {
+      String regexString = patternString.replace(".", "\\.");
+      regexString = regexString.replace("*", "([a-zA-Z\\.]*)");
+      excludes.add(Pattern.compile(regexString));
+    });
```
Comment (lanzone-myghty) [inline gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java]: Hey, supported syntax would be * in trailing position only. I will enhance it to work with the other valid java identifiers.

File: gatling-maven-plugin/pom.xml [hunk 2]
```diff
@@ -47,6 +49,18 @@
...
+			<groupId>org.powermock</groupId>
+			<artifactId>powermock-module-junit4</artifactId>
+			<version>${powermock.version}</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.powermock</groupId>
+			<artifactId>powermock-api-mockito</artifactId>
+			<version>${powermock.version}</version>
+			<scope>test</scope>
```
Comment (lanzone-myghty) [inline gatling-maven-plugin/pom.xml]: Yes, I reworked the stuff quite a few times so I did not see the clear solution anymore. Will introduce the getters and setters required for the variables and then it should be possible to work without powermock

**Comment by slandelle** [comment]
@lanzone-myghty I'm very sorry, I did review a few weeks ago, but forgot to validate/publish :(

File: gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java (hunk 1 above)
Comment (slandelle) [inline gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java]: Maybe have a look at how surefire does this: http://maven.apache.org/surefire/maven-surefire-plugin/examples/inclusion-exclusion.html

**Comment by lanzone-myghty** [comment]
no worries, have been busy anyway :) And I could have spotted the points you mention by myself. I also have to do the documentation. Just to many things at once going on, but I bet you know that

File: gatling-maven-plugin/pom.xml [hunk 3]
```diff
@@ -13,6 +13,8 @@
 	<packaging>maven-plugin</packaging>
 
 	<properties>
+		<powermock.version>1.6.6</powermock.version>
```
Comment (slandelle) [inline gatling-maven-plugin/pom.xml]: Please remove :)

File: gatling-maven-plugin/pom.xml [hunk 4]
```diff
@@ -99,4 +101,4 @@
 			</plugin>
 		</plugins>
 	</build>
-</project>
+</project>
```
Comment (slandelle) [inline gatling-maven-plugin/pom.xml]: Please restore last empty line so git is happy. Basically, you can revert all your changes on this file.

File: gatling-maven-plugin/src/test/java/io/gatling/mojo/WhitelistTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    String[] classFileList = new String[]{"io.test.great.class","io.test.great.work.class", "io.anothertest.awesome.class","io.onemoretest.cool.class"};
+    String[] includesList = new String[] {"io.test.great"};
+    GatlingMojo mojo = new GatlingMojo();
+
+    mojo.setIncludes(includesList);
+
+    List<String> result = mojo.resolveIncludesAndExcludes(classFileList);
+    Assert.assertTrue(result.size() == 1);
+  }
+}
```
Comment (slandelle) [inline gatling-maven-plugin/src/test/java/io/gatling/mojo/WhitelistTest.java]: Please add last empty line so git is happy

File: gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java (hunk 1 above)
Comment (slandelle) [inline gatling-maven-plugin/src/main/java/io/gatling/mojo/GatlingMojo.java]: Note that `java.lang.Character` has several method to check if a Character is valid in a Java identifier.

**Comment by alextotheroh** [comment]
Any updates on this?  It would greatly simplify things for us.  Thanks!

**Comment by slandelle** [comment]
@alextotheroh This PR can't be merged.

First, it conflicts.
Then, sadly, the filters implementation in no good. @lanzone-myghty tried to implement kind of ant style patterns, but those are broken. We can't pull an implementation that would work in most cases and unexpectedly fail otherwise.

My take on this is ant style patterns are complicated to implement properly (that's the reason why we dropped them in the Recorder and only support regular expressions).
If you're interested on contributing this, my advice is to build on top of this PR but go with regex instead of ant patterns.

**Comment by slandelle** [comment]
Feature implemented in d87e277

**Pull request closed without merging**